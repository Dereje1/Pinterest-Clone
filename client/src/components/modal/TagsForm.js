import React, { useState } from 'react';
import PropTypes from 'prop-types';
import DoneIcon from '@mui/icons-material/Done';
import TextField from '@mui/material/TextField';
import Autocomplete from '@mui/material/Autocomplete';

function TagsForm({
  closeTagsForm,
  addTag,
  suggestedTags,
  exisitingTags,
}) {
  const [tag, setTag] = useState('');

  // args for mui onInputChange callback => event, value, reason
  const handleTag = (_, value) => {
    if (value.trim().length < 16) {
      setTag(value);
    }
  };

  const handleTagSubmit = () => {
    const updatedTag = tag.trim().toUpperCase();
    if (updatedTag.length && !exisitingTags.includes(updatedTag)) {
      addTag(updatedTag);
    } else {
      setTag('');
    }
  };

  const handleEnterKey = (e) => {
    if (e.key === 'Enter') {
      handleTagSubmit();
    }
  };

  return (
    <div style={{
      width: 200,
      marginLeft: 'auto',
      marginRight: 'auto',
      display: 'flex',
      alignItems: 'flex-end',
    }}
    >
      <Autocomplete
        id="free-solo-tags"
        freeSolo
        options={suggestedTags.map((option) => option.toUpperCase())}
        fullWidth
        disableClearable
        inputValue={tag}
        autoComplete
        onInputChange={handleTag}
        onBlur={closeTagsForm}
        onKeyDown={handleEnterKey}
        filterOptions={(options) => {
          const filtered = [];
          options.forEach((o) => {
            if (!exisitingTags.includes(o) && o.startsWith(tag.toUpperCase())) {
              filtered.push(o);
            }
          });
          return filtered;
        }}
        renderInput={(params) => (
          <TextField
            {...params}
            autoFocus
            id="Tags_form"
            label="Add a Tag"
            type="text"
            variant="standard"
          />
        )}
      />

      <DoneIcon
        color="success"
        onClick={() => handleTagSubmit(tag)}
        onMouseDown={(e) => e.preventDefault()}
        style={{ cursor: 'pointer' }}
      />
    </div>
  );
}

export default TagsForm;

TagsForm.propTypes = {
  closeTagsForm: PropTypes.func.isRequired,
  addTag: PropTypes.func.isRequired,
  suggestedTags: PropTypes.arrayOf(PropTypes.string).isRequired,
  exisitingTags: PropTypes.arrayOf(PropTypes.string).isRequired,
};
