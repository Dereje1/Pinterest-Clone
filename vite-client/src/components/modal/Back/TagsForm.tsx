import React, { useState } from 'react';
import DoneIcon from '@mui/icons-material/Done';
import TextField from '@mui/material/TextField';
import Autocomplete from '@mui/material/Autocomplete';

interface TagsFormProps {
  closeTagsForm: () => void
  addTag: (tag: string) => void
  suggestedTags: string[]
  exisitingTags: string[]
}

function TagsForm({
  closeTagsForm,
  addTag,
  suggestedTags,
  exisitingTags,
}: TagsFormProps) {
  const [tag, setTag] = useState('');

  // args for mui onInputChange callback => event, value, reason
  const handleTag = (_: React.SyntheticEvent, value: string) => {
    if (value.trim().length < 16) {
      setTag(value);
    }
  };

  const handleTagSubmit = () => {
    const updatedTag = tag.trim().toUpperCase();
    if (updatedTag.length && !exisitingTags.includes(updatedTag)) {
      addTag(updatedTag);
    } else {
      setTag('');
    }
  };

  const handleEnterKey = (e: React.KeyboardEvent) => {
    if (e.key === 'Enter') {
      handleTagSubmit();
    }
  };

  return (
    <div style={{
      width: 200,
      marginLeft: 'auto',
      marginRight: 'auto',
      display: 'flex',
      alignItems: 'flex-end',
    }}
    >
      <Autocomplete
        id="free-solo-tags"
        freeSolo
        options={suggestedTags.map((option) => option.toUpperCase())}
        fullWidth
        disableClearable
        inputValue={tag}
        autoComplete
        onInputChange={handleTag}
        onBlur={closeTagsForm}
        onKeyDown={handleEnterKey}
        filterOptions={(options) => options.filter(
          (option) => !exisitingTags.includes(option) && option.startsWith(tag.toUpperCase()),
        )}
        renderInput={(params) => (
          <TextField
            {...params}
            autoFocus
            id="Tags_form"
            label="Add a Tag"
            type="text"
            variant="standard"
          />
        )}
      />

      <DoneIcon
        color="success"
        onClick={() => handleTagSubmit()}
        onMouseDown={(e) => e.preventDefault()}
        style={{ cursor: 'pointer' }}
      />
    </div>
  );
}

export default TagsForm;
