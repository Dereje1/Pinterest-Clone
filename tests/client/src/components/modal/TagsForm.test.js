import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import TagsForm from '../../../../../client/src/components/modal/TagsForm';

describe('The tags form component', () => {
  let props;
  beforeEach(() => {
    props = {
      addTag: jest.fn(),
      closeTagsForm: jest.fn(),
    };
  });

  afterEach(() => {
    props = null;
  });

  test('will render', () => {
    const wrapper = shallow(<TagsForm {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will update the value for characters <= 15', () => {
    const wrapper = shallow(<TagsForm {...props} />);
    let textField = wrapper.find('ForwardRef(TextField)');
    expect(textField.props().value).toBe('');
    // trigger value change
    textField.props().onChange({ target: { value: '15 characters..' } });
    textField = wrapper.find('ForwardRef(TextField)');
    expect(textField.props().value).toBe('15 characters..');
  });

  test('will not update the value for characters > 15', () => {
    const wrapper = shallow(<TagsForm {...props} />);
    let textField = wrapper.find('ForwardRef(TextField)');
    expect(textField.props().value).toBe('');
    // trigger value change
    textField.props().onChange({ target: { value: '16 characters..and more' } });
    textField = wrapper.find('ForwardRef(TextField)');
    expect(textField.props().value).toBe('');
  });

  test('will submit on done if value is present', () => {
    const wrapper = shallow(<TagsForm {...props} />);
    const textField = wrapper.find('ForwardRef(TextField)');
    // // trigger value change
    textField.props().onChange({ target: { value: '15 characters..' } });
    const done = wrapper.find('Memo(ForwardRef(DoneIcon))');
    // trigger submit
    done.props().onMouseDown({ preventDefault: jest.fn() });
    done.props().onClick();
    expect(props.addTag).toHaveBeenCalledWith('15 characters..');
  });

  test('will submit on enter key if value is present', () => {
    const wrapper = shallow(<TagsForm {...props} />);
    let textField = wrapper.find('ForwardRef(TextField)');
    // // trigger value change
    textField.props().onChange({ target: { value: '15 characters..' } });
    // trigger enter
    textField = wrapper.find('ForwardRef(TextField)');
    textField.props().onKeyDown({ key: 'Enter' });
    expect(props.addTag).toHaveBeenCalledWith('15 characters..');
  });

  test('will not submit on done if value is empty', () => {
    const wrapper = shallow(<TagsForm {...props} />);
    const textField = wrapper.find('ForwardRef(TextField)');
    // // trigger value change
    textField.props().onChange({ target: { value: '   ' } });
    const done = wrapper.find('Memo(ForwardRef(DoneIcon))');
    // trigger submit
    done.props().onMouseDown({ preventDefault: jest.fn() });
    done.props().onClick();
    expect(props.addTag).not.toHaveBeenCalled();
  });

  test('will not submit on any other key even if value is present', () => {
    const wrapper = shallow(<TagsForm {...props} />);
    let textField = wrapper.find('ForwardRef(TextField)');
    // // trigger value change
    textField.props().onChange({ target: { value: '15 characters..' } });
    // trigger enter
    textField = wrapper.find('ForwardRef(TextField)');
    textField.props().onKeyDown({ key: 'other key' });
    expect(props.addTag).not.toHaveBeenCalled();
  });
});
