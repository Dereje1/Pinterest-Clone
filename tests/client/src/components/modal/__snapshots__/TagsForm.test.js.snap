// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The tags form component will render 1`] = `
<div
  style={
    Object {
      "alignItems": "flex-end",
      "display": "flex",
      "marginLeft": "auto",
      "marginRight": "auto",
      "width": 200,
    }
  }
>
  <ForwardRef(Autocomplete)
    autoComplete={true}
    disableClearable={true}
    filterOptions={[Function]}
    freeSolo={true}
    fullWidth={true}
    id="free-solo-tags"
    inputValue=""
    onBlur={[MockFunction]}
    onInputChange={[Function]}
    onKeyDown={[Function]}
    options={
      Array [
        "SUGGESTED TAG 1",
        "SUGGESTED TAG 2",
      ]
    }
    renderInput={[Function]}
  />
  <Memo(ForwardRef(DoneIcon))
    color="success"
    onClick={[Function]}
    onMouseDown={[Function]}
    style={
      Object {
        "cursor": "pointer",
      }
    }
  />
</div>
`;
